### PR TITLE
For #14722：Support only one scaling job could be executed at the same time for one schema

### LINKED ISSUE #14722: Only one scaling job could be executed at the same time for one schema
### Is your feature request related to a problem?
No

### Describe the feature you would like.
In case of there're several scaling jobs are created that belong to the same schema, they will run concurrently, it might cause issue. So it's better to add some limitation:
1. Make sure there's only one scaling job is running if there're seveval scaling jobs that belong to the same schema.
2. Validate metadata, if target metadata is old than current enabled one, then it should not be executed. (It will be done after metadata multi-version is completed, not in this feature request.)

### PR DESCRIPTION
**Pull request by ReyYang**
Fixes #14722.

Changes proposed in this pull request:
- Add distributed lock
- Change Zookeeper distributed lock implementation
-

### CODE REVIEW

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java [hunk 1]
```diff
@@ -98,6 +108,10 @@ public void scalingTaskFinished(final ScalingTaskFinishedEvent event) {
             log.info("start to delete cache, ruleCacheId={}", ruleCacheId);
             registryCacheManager.deleteCache(SchemaMetaDataNode.getRulePath(event.getTargetSchemaName()), ruleCacheId);
         }
+        if (lockStatus) {
```
Comment (sandynz) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java:111]: From code, seems `ScalingTaskFinishedEvent` is not sent in every case, e.g. target DB broken and scaling job failed. The lock might be not released properly.

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java (hunk 1 above)
Comment (ReyYang) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java]: Yes, there are lock release failures when the task is stopped abruptly. Here I will raise a new pull request to solve it

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java [hunk 2]
```diff
@@ -98,6 +111,10 @@ public void scalingTaskFinished(final ScalingTaskFinishedEvent event) {
             log.info("start to delete cache, ruleCacheId={}", ruleCacheId);
             registryCacheManager.deleteCache(SchemaMetaDataNode.getRulePath(event.getTargetSchemaName()), ruleCacheId);
         }
+        if (schemaNameLockedMap.getOrDefault(decorateLockName(event.getTargetSchemaName()), false)) {
+            log.info("scaling job finished, release schema name lock, event = {}", event);
+            lockRegistryService.releaseLock(event.getTargetSchemaName());
```
Comment (sandynz) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java:116]: Seems `schemaNameLockedMap` just has `put` operation, but no `remove` operation, will it cause issue?

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java (hunk 1 above)
Comment (sandynz) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java]: OK, then you could comment it in this PR, it's clear to see it later.

**Comment by ReyYang** [comment]
There are lock release failures when the task is stopped abruptly. Here I will raise a new pull request to solve it

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java [hunk 3]
```diff
@@ -130,6 +127,20 @@ public void clusterSwitchConfiguration(final ClusterSwitchConfigurationEvent eve
...
+     * @param event Scaling Job Release Schema Name Lock Event
+     */
+    @Subscribe
+    public void scalingJobReleaseLockEvent(final ScalingJobReleaseSchemaNameLockEvent event) {
+        if (schemaNameLockedMap.getOrDefault(event.getSchemaName(), false)) {
+            log.info("scaling job finished, release schema name lock, event = {}", event);
+            lockRegistryService.releaseLock(decorateLockName(event.getSchemaName()));
+            schemaNameLockedMap.remove(event.getSchemaName());
+        }
+    }
```
Comment (sandynz) [inline shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java:142]: Is new event necessary for success case? Could we just do it in `scalingTaskFinished` like before, and add new logic for failure case later?

**Review by sandynz — CHANGES_REQUESTED** [review]

File: shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java (hunk 3 above)
Comment (ReyYang) [reply shardingsphere-mode/shardingsphere-mode-type/shardingsphere-cluster-mode/shardingsphere-cluster-mode-core/src/main/java/org/apache/shardingsphere/mode/manager/cluster/coordinator/registry/cache/subscriber/ScalingRegistrySubscriber.java]: ok, I will implement this feature in a new pull request

**Review by sandynz — APPROVED** [review]

**Pull request merged**